mlir/lib/Analysis/Presburger/Parser/ParseStructs.h

```diff
@@ -0,0 +1,345 @@
+//===- ParseStructs.h - Presburger Parse Structrures ------------*- C++ -*-===//
+//
+// Part of the LLVM Project, under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+//
+//===----------------------------------------------------------------------===//
+
+#ifndef MLIR_ANALYSIS_PRESBURGER_PARSER_PARSESTRUCTS_H
+#define MLIR_ANALYSIS_PRESBURGER_PARSER_PARSESTRUCTS_H
+
+#include "mlir/Analysis/Presburger/IntegerRelation.h"
+#include "mlir/Analysis/Presburger/PresburgerSpace.h"
+#include "llvm/ADT/ArrayRef.h"
+#include "llvm/ADT/STLExtras.h"
+#include "llvm/ADT/SmallVector.h"
+#include <cstdint>
+
+namespace mlir::presburger {
+using llvm::ArrayRef;
+using llvm::SmallVector;
+using llvm::SmallVectorImpl;
+
+/// This structure is central to the parser and flattener, and holds the number
+/// of dimensions, symbols, locals, and the constant term.
+struct ParseInfo {
+  unsigned numDims = 0;
+  unsigned numSymbols = 0;
+  unsigned numExprs = 0;
+  unsigned numDivs = 0;
+
+  constexpr unsigned getDimStartIdx() const { return 0; }
+  constexpr unsigned getSymbolStartIdx() const { return numDims; }
+  constexpr unsigned getLocalVarStartIdx() const {
+    return numDims + numSymbols;
+  }
+  constexpr unsigned getNumCols() const {
+    return numDims + numSymbols + numDivs + 1;
+  }
+  constexpr unsigned getConstantIdx() const { return getNumCols() - 1; }
+
+  constexpr bool isDimIdx(unsigned i) const { return i < getSymbolStartIdx(); }
+  constexpr bool isSymbolIdx(unsigned i) const {
+    return i >= getSymbolStartIdx() && i < getLocalVarStartIdx();
+  }
+  constexpr bool isLocalVarIdx(unsigned i) const {
+    return i >= getLocalVarStartIdx() && i < getConstantIdx();
+  }
+  constexpr bool isConstantIdx(unsigned i) const {
+    return i == getConstantIdx();
+  }
+};
+
+/// Helper for storing coefficients in canonical form: dims followed by symbols,
+/// followed by locals, and finally the constant term.
+///
+/// (x, y)[a, b]: y * 91 + x + 3 * a + 7
+/// coefficients: [1, 91, 3, 0, 7]
+struct CoefficientVector {
```

Comment:
Isn't this same as a IntegerRelation?

Reply:
No, an IntegerRelation holds IntMatrices of equalities and inequalities, and this structure is unsuitable for the parser, which parses one equality or inequality at a time. It is wasteful to re-use this data structure, filling in coefficients one column at a time, and resizing it every time there's a new column. SmallVector is the right data structure to use when one column is known at a time.

Reply:
I see, thanks for pointing that, i am indeed wrong.

It looks very similar to MultiAffineFunction. Please either use MultiAffineFunction or upstream AffineFunction to Presburger. I'm not okay with this being just a parser construct. I agree it's easier to land, but this will take a long time to refactor out. I would rather have this done right, than refactor things later.

Reply:
Regarding resizing, this parser is only used for tests. For anything worthy of being optimized, a user should directly construct what they want. I would prefer simplicity over optimizations for now.

Reply:
I'm not talking about some micro-optimization: I'm talking about using the right data structure for the task. It's poor engineering to use an unsuitable data structure, period. I have specialized APIs in CoefficientVector to make my parser code look nice, and these depend on iterating over a SmallVector.